README.md
# Kaggle - TGS Salt Identification Challenge

This repository contains my solution of the [Kaggle - TGS Salt Identification Challenge](https://www.kaggle.com/c/tgs-salt-identification-challenge). This competition is aimed to perform an Image Segmentation to detect where salt deposits are located in seismic images. The goal of the challenge is to optimize the [IoU metric](https://www.kaggle.com/c/tgs-salt-identification-challenge#evaluation) of the predicted maks.

The baseline model built consists in a [U-net](https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/) which takes a seismic image as input and outputs the predicted segmentation that indicates where the salt deposit is located.

<p>
  <img src="readme_images/input.png" width="200" title="Input image">
  <img src="readme_images/mask.png" width="200" title="Output mask">
  
  <em>Train input-output sample</em>
</p>

As the result of the competition, I have achieved an IoU value of 0.801 in the Public Leaderboard (1522/3291 position) and 0.824 in the Private Leaderboard (1479/3291 position), just 0.072 from the best solution.


## What worked

After performing a bunch of tests, in general terms, the bigger the Neural Network is, the more accurate is the prediction. But this size is limited by the hardware restrictions. 

Using **residual blocks** in the U-net layers leads to a more accurate prediction. The final prediction uses a [U-net](https://github.com/AlbertoSabater/Kaggle-Salt-Identification-Challenge/blob/master/nn_models.py#L177) built by 18 residual blocks, each of them made by 2 Convolutional layers. The model also uses BatchNormalization and Dropout to improve its performance. The NN also counts with 4 levels of MaxPooling in the encoder which are reverted with upsampling in the decoder. The padding of each Convolution layer is truncated to allow the NN to take the real input size (101 x 101 pixels) without the need of an image resizing, this improves the model performance and training time.

Since the training dataset is not too big, **Image Data Augmentation** is essential to not overfit the data. This augmentation involves horizontal flip, image rotation, zoom and width and height shift.

The [loss](https://github.com/AlbertoSabater/Kaggle-Salt-Identification-Challenge/blob/master/nn_models.py#L31) used to train the model uses a combination of the **Binary Cross-Entropy** and the **Dice Coefficient**.

Due to the size of the dataset and its data distribution, a **training with the full augmented dataset** and using the full no-augmented dataset as a test set performs better than a learning with train-test split.

The **threshold** used to get the final prediction is tuned to optimize the IoU metric used to evaluate the competition submission.

The final prediction is optimized by applying **Test-Time Augmentation** (TTA). With this method, each test sample is predicted several times with different Image Augmentations (horizontal flip in this case) and finally, each set of predictions is averaged to get a final result.

**Model Stacking** also improved prediction accuracy. In this way, an array of models (with the same architecture) is trained and the final prediction is averaged with all of their predictions.

To optimize the training time and performance a set of **[callbacks](https://github.com/AlbertoSabater/Kaggle-Salt-Identification-Challenge/blob/master/main.py#L227)** are added to it. 


## What didn't work

Train-test split didn't perform better on the leaderboard as a training with the full augmented dataset. Stratified train-test split (on salt coverage) and stratified k-fold have also been implemented but these techniques didn't show any improvement either.

Binary Cross-Entropy loss function performed worst that the combination of BCE and Dice.

The depth of each sample has been [added](https://github.com/AlbertoSabater/Kaggle-Salt-Identification-Challenge/blob/master/nn_models.py#L73) to the Neural Network as an array concatenated to the output of several convolutional layers with no improvement of the final performance.


## Work to do

The use of **pretrained encoders** (ResNet18, ResNet50, etc.) has been implemented with a learning that combines a training with a frozen encoder to unfroze it for a second training. However, due to the lack of GPU memory, this training couldn't be possible.

A **bigger batch size** would also improve the model performance but it hasn't been possible either due to the GPU restrictions.

A training with the **Lovasz Loss** or a combination of it with other losses would also have been improved the model performance.

User **[Hypercolumns](https://arxiv.org/pdf/1411.5752.pdf) to get the final prediction**.


CHANGELOG.md
# CHANGELOG

### 2018-10-19 | LB 1604/3243 | Score 0.793

Added K-fold training. Training a set of models with diferent train-test stratified splits. Final prediction based on the average of single model predictions  


### 2018-10-18 | LB 1604/3243 | Score 0.793

Test Time Augmentation (TTA) added to model predictions  
Train an array of models. Final prediction obtained with the average of single predictions  
IoU metric added to epoch metrics  
Added ReduceLROnPlateau to training callbacks  
Added pretrained u-net models based on ResNet-18 and ResNet-34. Unable to train them due to the lack of GPU memory  


### 2018-10-10 | LB 1781/3187 | Score 0.779

Ubuntu updated to Cuda 9.0  
Depth added to U-net concatenating it to first and/or mid layer  
Model store name updated  
Prediction threshold tunned with IoU metric  
Added more augmentation parameters  
U-net layer sizes configurable  
NEW MODEL, U-net with residual blocks created + configurable layer sizes. Not need image resampling  
Stratified train-test split on salt coverage  


### 2018-09-08 | LB 1870/2152 | Score 0.622

Dice Loss added  
Dice + BCE Loss added  
Dropout and BatchNormalization added to the baseline model  
Prediction threshold tunned  
Loading best weights after training  
Predictions compressed before submission  


### 2018-08-25 | LB 1321/1497 | Score 0.552

RLE prediction fixed  
Training without splitting train dataset into train and validation. Using train dataset without augmentation as validation dataset  
Added Windows support and its environment yml  
Parallel processing not working in Windows  
Preprocessed data is load or stored on disk for a quicker execution  
Train and test predicted dataset is stored on disk  
Submitting RLE predictions to Kaggle  
Using model_params as a comment


### 2018-08-23 | LB 1344/1432 | Score 0.347

Basic learning workflow implementation
* Data Loading and preprocessing in parallel
* Simple Data Augmentation in memory or on the fly
* Split train and validation data
* Load and store NN and architecture
* Fit NN
* Metric: binary_crossentropy
* Callbacks: ModelCheckpoint, EarlyStopping, TensorBoard, CSVLogger
	
Prediction + RLE + CSV submission file. Processing in parallel


nn_utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from keras.preprocessing.image import ImageDataGenerator
from joblib import Parallel, delayed 

from skimage.transform import resize
from keras.preprocessing.image import load_img

from tqdm import tqdm
import numpy as np
import os
import datetime
import sys
import platform
import pickle 
import re


BASE_MODEL_DIR = './models/'
BASE_PREPROCESSED_DATA_DIR = './data_preprocessed/'
if not os.path.exists(BASE_MODEL_DIR): os.makedirs(BASE_MODEL_DIR)
if not os.path.exists(BASE_PREPROCESSED_DATA_DIR): os.makedirs(BASE_PREPROCESSED_DATA_DIR)

N_JOBS = 1 if platform.system() == 'Windows' else 8


def create_new_model_folder():
	folder_num = len(os.listdir(BASE_MODEL_DIR))
	folder_path = BASE_MODEL_DIR + '{}_model_{}/'.format(datetime.datetime.today().strftime('%m%d_%H%M'), folder_num)
	os.mkdir(folder_path)
	return folder_path


# deprecated
def get_train_generator(batch_size, target_size):

	seed = 1
	
	data_gen_args = dict(rescale = 1./255)
	
	image_datagen = ImageDataGenerator(**data_gen_args)
	mask_datagen = ImageDataGenerator(**data_gen_args)
	
	image_generator = image_datagen.flow_from_directory(
		'./data/train/images',
		target_size=target_size,
		color_mode = 'grayscale',
		class_mode=None,
		batch_size = batch_size,
		seed=seed)
	
	mask_generator = mask_datagen.flow_from_directory(
		'./data/train/masks',
		target_size=target_size,
		color_mode = 'grayscale',
		class_mode=None,
		batch_size = batch_size,
		seed=seed)
	
	# combine generators into one which yields image and masks
	train_generator = zip(image_generator, mask_generator)
	
	return train_generator, image_datagen.n


def load_image(f, folder_path): return np.array(load_img(folder_path + f, color_mode='grayscale')) / 255
def process_image(img, target_size): 
	if img.shape == target_size: return img
	else: return resize(img, target_size+(1,), mode='constant', preserve_range=True, anti_aliasing=True)
def load_and_process_image(f, folder_path, target_size):
	img = load_image(f, folder_path)
	return process_image(img, target_size)

def load_folder_images(folder_path, target_size):
	pickle_file = '{}{}_{}_{}.pckl'.format(BASE_PREPROCESSED_DATA_DIR, 
				folder_path.split('/')[2], folder_path.split('/')[3], target_size)
	if os.path.isfile(pickle_file):
		print(' * Loading images from pickle: {}'.format(pickle_file))
		data, image_names = pickle.load(open(pickle_file, 'rb'))
		return data, image_names
	else: 	
		print(' * Loading and storing images: {}'.format(pickle_file))
		image_names = os.listdir(folder_path)
		images = Parallel(n_jobs=N_JOBS)(delayed(load_and_process_image)(f, folder_path, target_size) for f in tqdm(image_names, total=len(image_names)))
		data = np.array(images)
		pickle.dump((data, image_names), open(pickle_file, 'wb'))
		return data, image_names


# %%
		
def get_coverage_class(masks):
	mask_cov = [ (100*x.sum()/len(x.ravel())//10) for x in masks ]
#	mask_cov = masks.map(lambda x: x.sum()/len(x.ravel()))
	return mask_cov

#a = np.array([[[1,1,0,0,1], [0,1,0,1,1]],
#			  [[1,1,0,0,1], [0,1,1,1,1]]])
#get_coverage_class(full_train_masks)

# %%


#def get_image_generator_on_memory(images, masks, batch_size, data_gen_args):
#	
#	seed = 123
#
#	image_datagen = ImageDataGenerator(**data_gen_args)
#	mask_datagen = ImageDataGenerator(**data_gen_args)
#	
#	generator = zip(image_datagen.flow(images, batch_size=batch_size, shuffle=True, seed=seed), 
#						   mask_datagen.flow(masks, batch_size=batch_size, shuffle=True, seed=seed))
#	
#	return generator, images.shape[0]


#def get_image_depth_generator_on_memory(images, masks, depths, batch_size, data_gen_args):
#	
#	seed = 123
#
#	image_datagen = ImageDataGenerator(**data_gen_args)
#	mask_datagen = ImageDataGenerator(**data_gen_args)
##	depth_datagen = ImageDataGenerator({})
#	
#	generator = zip(image_datagen.flow(images, batch_size=batch_size, shuffle=True, seed=seed),
##					   depth_datagen.flow(depths, batch_size=batch_size, shuffle=True, seed=seed), 
#					   mask_datagen.flow(masks, batch_size=batch_size, shuffle=True, seed=seed))
#	
#	return generator, images.shape[0]


def get_image_generator_on_memory_v2(images, masks, batch_size, data_gen_args):
	seed = 123
	image_datagen = ImageDataGenerator(**data_gen_args)
	mask_datagen = ImageDataGenerator(**data_gen_args)	
	
	image_f = image_datagen.flow(images, batch_size=batch_size, shuffle=True, seed=seed)
	mask_f = mask_datagen.flow(masks, batch_size=batch_size, shuffle=True, seed=seed)
	
	while True:
		image_n = image_f.next()
		mask_n = mask_f.next()
		yield image_n, mask_n
		
def get_image_depth_generator_on_memory_v2(images, masks, depths, batch_size, data_gen_args):
	seed = 123
	image_datagen = ImageDataGenerator(**data_gen_args)
	mask_datagen = ImageDataGenerator(**data_gen_args)	
	
	image_f = image_datagen.flow(images, depths, batch_size=batch_size, shuffle=True, seed=seed)
	mask_f = mask_datagen.flow(masks, batch_size=batch_size, shuffle=True, seed=seed)
	
	while True:
		image_n = image_f.next()
		mask_n = mask_f.next()
		yield [image_n[0], image_n[1]], mask_n



# deprecated
def rle(img, base_score):
	
	flat_img = img.flatten()
	flat_img = np.where(flat_img > base_score, 1, 0).astype(np.uint8)
	flat_img = np.insert(flat_img, [0, len(flat_img)], [0, 0])

	starts = np.array((flat_img[:-1] == 0) & (flat_img[1:] == 1))
	ends = np.array((flat_img[:-1] == 1) & (flat_img[1:] == 0))
	starts_ix = np.where(starts)[0] + 1
	ends_ix = np.where(ends)[0] + 1
	lengths = ends_ix - starts_ix
	
	return ' '.join([ str(s)+' '+str(l) for s, l in zip(starts_ix, lengths) ])


## RLE encoding, as suggested by Tadeusz Hupało
def rle_encoding(x, base_score):
	x = np.where(x > base_score, 1, 0)
	dots = np.where(x.T.flatten() == 1)[0]
	run_lengths = []
	prev = -2
	for b in dots:
		if (b>prev+1): run_lengths.extend((b + 1, 0))
		run_lengths[-1] += 1
		prev = b
	return re.sub(r'[\[\],]','', str(run_lengths))


def mask_to_rle(mask, base_score):
	mask = np.where(mask > base_score, 1, 0)

	mask_flat = mask.flatten('F')
	flag = 0
	rle_list = list()
	for i in range(mask_flat.shape[0]):
		if flag == 0:
			if mask_flat[i] == 1:
				flag = 1
				starts = i+1
				rle_list.append(starts)
		else:
			if mask_flat[i] == 0:
				flag = 0
				ends = i
				rle_list.append(ends-starts+1)
	if flag == 1:
		ends = mask_flat.shape[0]
		rle_list.append(ends-starts+1)
	#sanity check
	if len(rle_list) % 2 != 0:
		print('NG')
	if len(rle_list) == 0:
		rle = np.nan
	else:
		rle = ' '.join(map(str,rle_list))
	return rle

def rle_to_mask(rle_list, SHAPE):
    tmp_flat = np.zeros(SHAPE[0]*SHAPE[1])
    if len(rle_list) == 1:
        mask = np.reshape(tmp_flat, SHAPE).T
    else:
        strt = rle_list[::2]
        length = rle_list[1::2]
        for i,v in zip(strt,length):
            tmp_flat[(int(i)-1):(int(i)-1)+int(v)] = 255
        mask = np.reshape(tmp_flat, SHAPE).T
    return mask
	

def get_result(pred, base_score):
	pred = resize(pred, (101, 101), mode='constant', preserve_range=True, anti_aliasing=True)
#	pred = rle(pred, base_score)
	pred = rle_encoding(pred, base_score)
#	pred = mask_to_rle(pred, base_score)
	return pred

#def get_prediction_result(model, images, target_size, base_score):
#	return Parallel(n_jobs=N_JOBS)(delayed(get_result)(model.predict(img.reshape((1,)+target_size+(1,)))[0,:,:,0], base_score) 
#						for img in tqdm(images, total=len(images)))


main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import nn_models
import nn_utils
import matplotlib.pyplot as plt

#from PIL import Image
import numpy as np
import pandas as pd
import os
import json
import datetime
import platform
import pickle
import time
#from tqdm import tqdm

from keras.callbacks import EarlyStopping, ModelCheckpoint, TensorBoard, CSVLogger, ReduceLROnPlateau
from keras.models import load_model, model_from_json
from sklearn.model_selection import train_test_split
from sklearn import preprocessing

import segmentation_models
from segmentation_models.utils import set_trainable
#from resnet_pretrained import UResNet34


TENSORBOARD_DIR = './logs/'


model_params = {
			'datetime': str(datetime.datetime.now()),
			'stratify': True,
			'n_models': 5,
			'model_type': 'rrunet', 		# resnet34 / rrunet
			'freeze_encoder': True,
			'retrain_resnet': True,
			'target_size': (101,101),
#			'target_size': (256,256),
			'nn_size_base': 16,
			'include_depth': [],
			'dropout': True,
			'batchnorm': True,
			'data_gen_args': {
					'horizontal_flip': True,
					'vertical_flip': True,
					'rotation_range': 15,
					'width_shift_range': 0.1,
					'height_shift_range': 0.1,
					'zoom_range': [0.9, 1.2]
				},
			'validation_split': 0.0,
			'batch_size': 32,
			'epochs': 250,
			'es_patience': 10,
			'rlr_patience': 5,																																																																												
			'loss': 'bcedice', 			# bce / -dice- / bcedice / -bceiou-
			'optimizer': 'adam',
			'metrics': ["accuracy", nn_models.iou_tf],
			'monitor': 'val_loss', 'monitor_mode': 'min',
			'model_architecture_file': 'model_architecture',
			'base_score': 0.5,
			'tta': True
		}

if not os.path.exists(TENSORBOARD_DIR): os.makedirs(TENSORBOARD_DIR)

depths = pd.DataFrame.from_csv('./data/depths.csv')
depths.z = preprocessing.MinMaxScaler().fit_transform(depths[['z']].values.astype(float))


# %%
	
# =============================================================================
# Create train data generator and validation set
# =============================================================================

# Load images
full_train_images, full_train_image_names = nn_utils.load_folder_images("./data/train/images/", model_params['target_size'])
full_train_masks, _ = nn_utils.load_folder_images("./data/train/masks/", model_params['target_size'])
coverage_class = nn_utils.get_coverage_class(full_train_masks)


# %%

if model_params['model_type'] in ['resnet34', 'resnet18']:
	full_train_images = np.array([ np.dstack([img]*3) for img in full_train_images ])


# %%

# Create train and validation dataset
if model_params['validation_split'] > 0:
	train_images, val_images, train_masks, val_masks= train_test_split(
			full_train_images, full_train_masks, test_size=model_params['validation_split'], random_state=123)
	if model_params['stratify'] and model_params['validation_split']>0.0:
		train_image_names, val_image_names = train_test_split(full_train_image_names, 
					test_size=model_params['validation_split'], stratify=coverage_class, random_state=123)
	else: 
		train_image_names, val_image_names = train_test_split(full_train_image_names, test_size=model_params['validation_split'], random_state=123)
else:
	train_images = val_images = full_train_images.copy()
	train_masks = val_masks = full_train_masks.copy()
	train_depths = val_depths = depths.copy()
	train_image_names = val_image_names = full_train_image_names.copy()


if model_params['model_type'] not in ['resnet34', 'resnet18']:
	train_images = train_images.reshape((len(train_images),)+model_params['target_size']+(1,))
	val_images = val_images.reshape((len(val_images),)+model_params['target_size']+(1,))
	
model_params['num_train'] = len(train_images)
model_params['num_val'] = len(val_images)

# Augmented train dataset
#train_generator, num_samples_train = get_train_generator(batch_size, target_size)
if len(model_params['include_depth']) > 0:
	
	print(' * Including depth')
	train_depths = depths.loc[[ n[:-4] for n in train_image_names ]].values
	val_depths = depths.loc[[ n[:-4] for n in val_image_names ]].values
#	depths_train = np.concatenate( [ np.resize(d, (1, 128,128,1)) for d in depths_train ])
#	train_generator, num_samples_train = nn_utils.get_image_depth_generator_on_memory(
	train_generator = nn_utils.get_image_depth_generator_on_memory_v2(
			train_images, train_masks, train_depths,
			model_params['batch_size'], model_params['data_gen_args'])
	model_params['num_samples_train'] = train_images.shape[0]
	
	train_data = [train_images, train_depths]
	val_data = [val_images, val_depths]
	
else:
	print(' * No including depth')
	train_generator = nn_utils.get_image_generator_on_memory_v2(
			train_images, train_masks,
			model_params['batch_size'], model_params['data_gen_args'])
	model_params['num_samples_train'] = train_images.shape[0]
	train_data = train_images
	val_data = val_images

print(' *  Data generator ready')


# %%
	
# =============================================================================
# Create and train Model
# =============================================================================

model_params['model_folder'] = nn_utils.create_new_model_folder()
if not os.path.exists(model_params['model_folder']): os.makedirs(model_params['model_folder'])
model_params['model_weights_file']= 'model_weights'

print(model_params['model_folder'])


models = []
for i in range(model_params['n_models']):
	
	print(' *  Building', model_params['model_type'])

	if model_params['model_type'] == 'unet':
		model = nn_models.unet(input_size = model_params['target_size'] + (1,),
							   size_base = model_params['nn_size_base'],
							   include_depth = model_params['include_depth'],
							   dropout=model_params['dropout'],
							   batchnorm=model_params['batchnorm'],)
	elif model_params['model_type'] == 'rrunet':
		model = nn_models.residual_unet(input_size = model_params['target_size'] + (1,),
								size_base = model_params['nn_size_base'],
								include_depth = model_params['include_depth'])
	elif model_params['model_type'] == 'resnet34':
		model = segmentation_models.Unet(backbone_name='resnet34', 
								   encoder_weights='imagenet', freeze_encoder=model_params['freeze_encoder'])
#		img_size_target = 128	
#		model = UResNet34(input_shape=(1,)+model_params['target_size'], freeze_encoder=True)
	elif model_params['model_type'] == 'resnet18':
		model = segmentation_models.Unet(backbone_name='resnet18', 
								   encoder_weights='imagenet', freeze_encoder=model_params['freeze_encoder'])
	else:
		raise ValueError('Model type Error')
			
		
#	print(model.summary())
	print('Layers unfrozen: {}/{}'.format(sum([l.trainable for l in model.layers]), len(model.layers)))
	
	
	nn_loss = 'binary_crossentropy'
	if model_params['loss'] == 'bce':
		nn_loss = nn_models.bce
	elif model_params['loss'] == 'dice':
		nn_loss = nn_models.dice_loss
	elif model_params['loss'] == 'bcedice':
		nn_loss = nn_models.bce_dice_loss
		
	model.compile(loss=nn_loss, optimizer=model_params['optimizer'], metrics=model_params['metrics'])
	
	model_checkpoint_file = model_params['model_folder'] + str(len(models)) + '_' + model_params['model_weights_file'] + '.h5'
	callbacks = [
				ModelCheckpoint(model_checkpoint_file, 
					   monitor=model_params['monitor'], verbose=1, save_best_only=True, mode=model_params['monitor_mode']),
					   
				EarlyStopping(monitor=model_params['monitor'], min_delta=0.00001, 
					 verbose=1, mode=model_params['monitor_mode'], patience=model_params['es_patience']),
				
				TensorBoard(log_dir='{}{}'.format(TENSORBOARD_DIR, model_params['model_folder'].split('/')[-2]), 
							  histogram_freq=0, write_graph=True, 
							  write_grads=1, batch_size=model_params['batch_size'], write_images=True),
				   
				CSVLogger(model_params['model_folder']+'log.csv', separator=',', append=True),
				
				ReduceLROnPlateau(monitor=model_params['monitor'], mode=model_params['monitor_mode'], 
								  factor=0.5, patience=model_params['rlr_patience'], 
								  min_lr=0.00001, verbose=1)
			]
	
	print(' *  Model {}/{} ready'.format(i+1, model_params['n_models']))



	hist = model.fit_generator(
				generator = train_generator,
				steps_per_epoch = 1.4*(model_params['num_samples_train'] // model_params['batch_size']), #################### 3072 num_samples_train
				epochs = model_params['epochs'],
				validation_data = (val_data, val_masks),
	#			validation_steps = num_samples_val // batch_size,
				shuffle = True,
				callbacks = callbacks,
				use_multiprocessing = False if platform.system() == 'Windows' else True,
				verbose = 1)


	if model_params['retrain_resnet'] and model_params['freeze_encoder'] and \
			model_params['model_type'] in ['resnet34', 'resnet18']:
		print(' * Unfreezing model')
#		model = nn_models.set_model_trainable(model)
#		model.compile(loss=nn_loss, optimizer=model_params['optimizer'], metrics=model_params['metrics'])
		set_trainable(model)
		print('Layers unfrozen: {}/{}'.format(sum([l.trainable for l in model.layers]), len(model.layers)))
		hist = model.fit_generator(
					generator = train_generator,
					steps_per_epoch = 1.4*(model_params['num_samples_train'] // model_params['batch_size']), #################### 3072 num_samples_train
					epochs = model_params['epochs'],
					validation_data = (val_data, val_masks),
		#			validation_steps = num_samples_val // batch_size,
					shuffle = True,
					callbacks = callbacks,
					use_multiprocessing = False if platform.system() == 'Windows' else True,
					verbose = 1)
	
	print(' *  Model {}/{} trained'.format(i+1, model_params['n_models']))
	model.load_weights(model_checkpoint_file)
	models.append(model)
	
	val_loss = min(hist.history['val_loss'])
	val_iou = min(hist.history['val_iou_tf'])
	os.rename(model_checkpoint_file, 
		   model_checkpoint_file.replace('.h5', '_{:.4f}_{:.4f}.h5'.format(val_loss, val_iou)))


# %%



# %%


# Store model architecture
model_architecture = model.to_json()
with open(model_params['model_folder'] + model_params['model_architecture_file'] + '.json', 'w') as f:
	json.dump(model_architecture, f)
	

del model_params['metrics']
# Store model params
with open(model_params['model_folder'] + 'model_params.json', 'w') as f:
	json.dump(model_params, f)


# Rename model_folder with monitor and its value, and tensorboard folder
val = min(hist.history[model_params['monitor']]) if 'loss' in model_params['monitor'] else max(hist.history[model_params['monitor']])
new_model_folder = model_params['model_folder'][:-1] + \
				"_{}_{:.4f}".format(model_params['monitor'], val)
new_model_folder = new_model_folder + '_{}_nm{}_{}_l{}_sb{}_{}_d{}_{}_{}/'.format(
						model_params['model_type'],
						model_params['n_models'],
						'stry' if model_params['stratify'] and model_params['validation_split']>0.0 else 'nstry',
						model_params['loss'],
						model_params['nn_size_base'],
						'd' if model_params['dropout'] else 'nd',
						'bn' if model_params['batchnorm'] else 'nbn',
						''.join([ str(v) for v in model_params['include_depth'] ]) if len(model_params['include_depth'])>0 else 'nd',
						'tta' if model_params['tta'] else 'notta'
		)

os.rename(model_params['model_folder'], new_model_folder)
os.rename(model_params['model_folder'].replace('./models/', './logs/'), new_model_folder.replace('./models/', './logs/'))
model_params['model_folder'] = new_model_folder


# %%

original_masks, _ = nn_utils.load_folder_images("./data/train/masks/", (101,101))
if model_params['validation_split'] > 0:
	train_original_masks, val_origial_masks = train_test_split(original_masks, test_size=model_params['validation_split'], random_state=123)
else:
	train_original_masks = val_origial_masks = original_masks
	
	
# %%

val_preds = nn_models.predict_models(models, val_data, tta=model_params['tta'])
#val_preds = model.predict(val_data)
val_preds = nn_utils.process_image(val_preds, (len(val_preds),)+(101,101))

# %%

print('Tuning base score', datetime.datetime.now().strftime('%H:%M:%S'))
t = time.time()
#best_bs, best_score = nn_models.tune_base_score(val_preds.ravel(), val_origial_masks.ravel().astype(int))
best_bs, best_score = nn_models.tune_base_score(val_preds, val_origial_masks.astype(int))
model_params['base_score'] = best_bs
print('Base score tuned. {}. {:.2f}m. {:.4f} | {}'.format(datetime.datetime.now().strftime('%H:%M:%S'), 
	  (time.time()-t)/60, best_bs, best_score))


# %%

print(' * Calculating and storing train predictions')
#train_preds = nn_utils.get_prediction_result(model, val_data, model_params['target_size'], model_params['base_score'])
#train_preds = model.predict(train_data)
train_preds = nn_models.predict_models(models, train_data, tta=model_params['tta'])
train_preds = [ nn_utils.get_result(tp, model_params['base_score']) for tp in train_preds ]
pickle.dump((train_preds, full_train_image_names), open(model_params['model_folder']+'preds_train.pckl', 'wb'))


# %%

# =============================================================================
# Predict data
# =============================================================================

test_dir = './data/test/'
test_data, test_image_names = nn_utils.load_folder_images(test_dir, model_params['target_size'])
test_data = test_data.reshape((len(test_data),)+model_params['target_size']+(1,))
if len(model_params['include_depth']) > 0:
	test_data = [test_data, depths.loc[[ n[:-4] for n in test_image_names ]].values]

	
print(' * Calculating and storing test predictions')

#test_preds = model.predict(test_data)
test_preds = nn_models.predict_models(models, test_data, tta=model_params['tta'])
test_preds = [ nn_utils.get_result(tp, model_params['base_score']) for tp in test_preds ]

pickle.dump((test_preds, test_image_names), open(model_params['model_folder']+'preds_test.pckl', 'wb'))

csv_df = pd.DataFrame.from_dict({'id': [ n.split('.')[0] for n in test_image_names ],
									'rle_mask': test_preds })

csv_path = model_params['model_folder']+model_params['model_folder'].split('/')[-2]+'.csv.gz'
csv_df.to_csv(csv_path, compression='gzip', index=False)


print(' ** Submit prediction?')
#input()
print(' * Submitting predictions')
#comment = '\n'.join([ '{}: {}'.format(k,v) for k,v in model_params.items() ])
#comment = str(model_params)

val_loss = max(hist.history['val_loss'])
comment = 'l{}_vl{:.4f}_vs{}'.format(model_params['loss'], val_loss, model_params['validation_split'])
command = 'kaggle competitions submit -c tgs-salt-identification-challenge -f {} -m "{}"'.format(csv_path.replace('|', '\|'), comment)
print(command)
os.system(command)
print(' * Predictions submitted')





# %%

#model.load_weights('./models/1004_0915_model_24_val_loss_0.1887_bce_d_bn_d1/model_weights.h5')


# %%

# Plot train results
if False:
# %%
#	inds = np.random.permutation(range(len(val_images)))
	while True:
		ind = np.random.choice(list(range(len(val_images))))
		
		img = val_images[ind].reshape((1,)+model_params['target_size']+(1,))
		mask = val_masks[ind].reshape((1,)+model_params['target_size']+(1,))

		if len(model_params['include_depth']) > 0:
			depth = val_depths[ind]
			pred = model.predict([img, depth])[0,:,:,0]
		else:
			pred = model.predict(img)[0,:,:,0]

		
		fig = plt.figure()
		ax = plt.subplot(221)
#		plt.imshow(i[0][0,:,:,0])
		plt.imshow(img[0,:,:,0])
		ax.set_title("Orig")
		ax = plt.subplot(222)
#		plt.imshow(i[1][0,:,:,0])
		plt.imshow(mask[0,:,:,0])
		ax.set_title("Real mask")
		ax = plt.subplot(223)
		plt.imshow(pred)
		ax.set_title("Pred")
		ax = plt.subplot(224)
		plt.imshow(np.where(pred > model_params['base_score'], 1, 0))
		ax.set_title("Pred mask")
		
		pred_res = nn_utils.resize(pred, (101, 101), mode='constant', preserve_range=True, anti_aliasing=True)
		rle = nn_utils.rle_encoding(pred_res, model_params['base_score'])
		pred_fin = nn_utils.rle_to_mask(rle.split(), (101,101))
		pred_send = nn_utils.get_result(pred, model_params['base_score'])
		pred_send = nn_utils.rle_to_mask(pred_send.split(), (101,101))
		plt.figure()
		plt.subplot(131)
		plt.imshow(pred_res)
		plt.subplot(132)
		plt.imshow(pred_fin)
		plt.subplot(133)
		plt.imshow(pred_send)
		
		break
	
	# %%
	
# Plot test_results
if False:
	
	# %%
#	test_dir = 'data/test/'
#	test_image_names = os.listdir(test_dir)
#	img = nn_utils.load_and_process_image(np.random.choice(test_image_names, 1)[0], test_dir, model_params['target_size'])
	
#	pred = model.predict(img.reshape((1,)+model_params['target_size']+(1,)))[0,:,:,0]
	ind = np.random.choice(list(range(len(val_images))))
	img = test_data[0][ind].reshape((1,)+model_params['target_size']+(1,))
	
	if len(model_params['include_depth']) > 0:
		depth = test_data[1][ind]
		pred = model.predict([img, depth])[0,:,:,0]
	else:
		pred = model.predict(img.reshape((1,)+model_params['target_size']+(1,)))[0,:,:,0]
	
	fig = plt.figure()
	plt.subplot(131)
	plt.imshow(img[0,:,:,0])
	plt.subplot(132)
	plt.imshow(pred)
	plt.subplot(133)
	plt.imshow(np.where(pred > model_params['base_score'], 1, 0))
	print(test_image_names[ind])
	
	
	# %%
	
# Plot test_results
if False:
	
	# %%
#	test_dir = 'data/test/'
#	test_image_names = os.listdir(test_dir)
#	img = nn_utils.load_and_process_image(np.random.choice(test_image_names, 1)[0], test_dir, model_params['target_size'])
	
#	pred = model.predict(img.reshape((1,)+model_params['target_size']+(1,)))[0,:,:,0]
	ind = np.random.choice(list(range(len(csv_df))))
	img_name = csv_df.iloc[ind].id
	img_rle = csv_df.iloc[ind].rle_mask
	img = nn_utils.rle_to_mask(img_rle.split(), (101,101))
	img_orig = test_data[0][test_image_names.index(img_name+'.png')]
	
#	img = test_data[0][ind].reshape((1,)+model_params['target_size']+(1,))
#	
#	if len(model_params['include_depth']) > 0:
#		depth = test_data[1][ind]
#		pred = model.predict([img, depth])[0,:,:,0]
#	else:
#		pred = model.predict(img.reshape((1,)+model_params['target_size']+(1,)))[0,:,:,0]
	
	fig = plt.figure()
	plt.subplot(131)
	plt.imshow(img_orig[:,:,0])
	plt.subplot(132)
	plt.imshow(img)
#	plt.subplot(133)
#	plt.imshow(np.where(pred > model_params['base_score'], 1, 0))
	print(img_name)
	

# %%
	
for r in train_generator:
	pass
	break
	
	